### PR TITLE
POCA trainer

### PR DESCRIPTION
**Pull request by andrewcoh**
### Proposed change(s)

This PR adds the POCA trainer and associated tests. In addition it makes changes to the extrinsic reward provider to enable team-based rewards to work. 

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)

PR for documentation - to be merged after this one #5056
[Explanation](https://docs.google.com/document/d/1TJHImYY1N3a1xh_YOEWoY2aO81CygHdfbLs3IgeoXGo/edit#) of some of the design choices:

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: config/ppo/PushBlockCollab.yaml
Comment (ervteng) [inline config/ppo/PushBlockCollab.yaml:1]: TODO: remove this file and move to Hunter's pushblock branch

File: ml-agents/mlagents/trainers/coma/trainer.py
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/coma/trainer.py:3]: remove

File: ml-agents/mlagents/trainers/coma/trainer.py
Comment (awjuliani) [inline ml-agents/mlagents/trainers/coma/trainer.py:25]: Presumably this name is temporary, until we are on the official one, right?

File: ml-agents/mlagents/trainers/coma/trainer.py
Comment (awjuliani) [inline ml-agents/mlagents/trainers/coma/trainer.py:26]: This will be a great spot for the arxiv link :D

File: ml-agents/mlagents/trainers/coma/trainer.py
Comment (awjuliani) [inline ml-agents/mlagents/trainers/coma/trainer.py:39]:
```suggestion
        Responsible for collecting experiences and training COMA model.
```

File: ml-agents/mlagents/trainers/coma/trainer.py
Comment (awjuliani) [inline ml-agents/mlagents/trainers/coma/trainer.py:301]: @andrewcoh Do you have a link to the equation this is implementing?

File: ml-agents/mlagents/trainers/settings.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/settings.py:221]: Is this a new hyperparameter for the user ?
Looks like the value is hard coded anyway in the coma trainer, is it really necessary ?

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:343]: I do not know how I feel about these two methods. I think they would look better as a module given how they are used.

File: ml-agents/mlagents/trainers/coma/trainer.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/coma/trainer.py:145]: GAE is basically the lambda return over n step advantages. It's possible we could reuse the `discount_rewards` function from the PPO trainer but I'm not sure it's worth putting that function in a new file just for this purpose.

File: ml-agents/mlagents/trainers/coma/trainer.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/coma/trainer.py:301]: This is a pretty good discussion of TD(lambda) I think. https://amreis.github.io/ml/reinf-learn/2017/11/02/reinforcement-learning-eligibility-traces.html

TD(lambda) was also used in COMA https://github.com/oxwhirl/pymarl/blob/master/src/utils/rl_utils.py#L4

File: ml-agents/mlagents/trainers/torch/networks.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/networks.py:118]: It's an artifact. removed.

File: ml-agents/mlagents/trainers/settings.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/settings.py:221]: It is not a new hyperparameter and it is actually to avoid requiring users to learn about a new group reward signal. POCA will set this to True on initialization so that the ExtrinsicReward provider knows to collect teammate rewards. Otherwise it will be False. The only way for users to change it is to modify python.

File: ml-agents/mlagents/trainers/torch/networks.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/networks.py]: Renamed to MultiAgentNetworkBody and added docstring.

I think you're right but that we should leave this for a refactor later on.

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/utils.py:343]: Maybe we can combine this with a change to NetworkBody to also satisfy the comment that MultiAgentNetworkBody should reuse NetworkBody. The right solution may be to rename NetworkBody to Encoders that has the functionality of these methods and move the linear layer to the Actor/Critic. Then the MultiAgentNetBody may be able to reuse the encoders.

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/utils.py:343]: reference https://github.com/Unity-Technologies/ml-agents/pull/5005#discussion_r591813207

File: ml-agents/mlagents/trainers/torch/networks.py
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/networks.py]: Why leave this for later ? It is not a refactor, it is a new feature. And I think we should be careful with adding new refactor work for later.

File: ml-agents/mlagents/trainers/settings.py
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/settings.py:221]: Then why is it a setting and not a constant in the code?

File: ml-agents/mlagents/trainers/torch/networks.py
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/networks.py]: I'm going to play devil's advocate here, and say it's probably not going to be a good idea to try and reuse networkbody. Originally I tried to do that during the prototyping phase and there were too many special cases. The dataflow is just different. 

A networkbody goes `obs -> encoders -> RSA -> linear layers -> output`. 
The MultiInput one has to do `obs -> concat with other obs -> encoders x num_obs -> RSA x num_obs -> another RSA -> linear layers -> output.`

So really what we're re-using is the RSA, the encoders, and the linear layer, but in a different order, and these are all already separate modules. To re-use NetworkBody we'd have to insert the 2nd RSA inside the first NetworkBody.

File: ml-agents/mlagents/trainers/settings.py
Comment (ervteng) [reply ml-agents/mlagents/trainers/settings.py:221]: The alternative would be to have a public variable that is set later (ew) or additional constructor parameter to Extrinsic Reward Provider (and only Extrinsic Reward Provider). This would mess with the BaseRewardProvider abstraction. I don't really have a preference, but it is choosing between the lesser of several evils here.

Comment (ervteng) [reply ml-agents/mlagents/trainers/settings.py:221]: Potential alternative: #5082

**Review by ervteng — APPROVED** [review]
LGTM, will wait for at least one more reviewer

File: ml-agents/mlagents/trainers/poca/optimizer_torch.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/poca/optimizer_torch.py:105]: Explain the meaning of the parameters (especially for `baseline` since it is not an interface implementation.

File: ml-agents/mlagents/trainers/poca/optimizer_torch.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/poca/optimizer_torch.py:119]: Remove this method. It has no reason to be public.

File: ml-agents/mlagents/trainers/poca/optimizer_torch.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/poca/optimizer_torch.py:240]: Some of my comments got lost. Please review them in the conversation tab : https://github.com/Unity-Technologies/ml-agents/pull/5005#discussion_r591824890

File: ml-agents/mlagents/trainers/torch/components/reward_providers/extrinsic_reward_provider.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/components/reward_providers/extrinsic_reward_provider.py:32]: Invert these 2 ifs. No need to check the first one if there are no groumaterewards

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/components/reward_providers/extrinsic_reward_provider.py:32]: These if conditions could be better:

if self.add_groupmate_rewards and BufferKey.GROUPMATE_REWARDS in mini_batch : do the groupmate reward
if BufferKey.GROUP_REWARD in mini_batch : Do the group reward

File: ml-agents/mlagents/trainers/torch/networks.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/networks.py:224]: I cannot help but think there is a better way. Something like 
obs_without_action: List[List[torch.Tensor]],
obs_action: List[Tuple[List[torch.Tensor], List[torch.Tensor]]]
I know the types are ugly, but it should be explicit that the length of obs and actions has to be equal

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:376]: Stand by my statement, make create_residual_self_attention a module with encode_observations its forward method

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/utils.py:376]: Call it ObservationEncoder

File: ml-agents/mlagents/trainers/settings.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/settings.py:609]: For clarity define POCA settings as 
```
POCASettings = PPOSettings
```
Will set us up better when the two settings diverge

**Review by vincentpierre — APPROVED** [review]
I approve once all comments have been resolved. 
Including those that were wrongfully marked as outdated by github.

File: ml-agents/mlagents/trainers/torch/components/reward_providers/extrinsic_reward_provider.py
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/components/reward_providers/extrinsic_reward_provider.py:32]: Updated

File: ml-agents/mlagents/trainers/torch/networks.py
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/networks.py:224]: I'm actually not in favor of having nested tuples as parameters, it doesn't enforce length to be equal any more than two parameters, and requires constructing additional datastructures before use.

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/networks.py:296]: https://github.com/Unity-Technologies/ml-agents/pull/5094/

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/utils.py]: courtesty of @ervteng https://github.com/Unity-Technologies/ml-agents/pull/5093/

**Pull request merged**
